graalpy/graalpy-scripts-debug-guide/README.md

```diff
@@ -0,0 +1,244 @@
+
+`App.java`
+```java
+public class App extends Application {
+    private Context context;
+
+    @Override
+    public void stop() throws Exception {
+        context.close();
+        super.stop();
+    }
+
+    @Override
+    public void start(Stage stage) {
+        context = Context.newBuilder("python")
+            .allowIO(IOAccess.newBuilder() // ①
+                            .fileSystem(FileSystem.newReadOnlyFileSystem(FileSystem.newDefaultFileSystem()))
+                            .build())
+            .allowPolyglotAccess(PolyglotAccess.newBuilder() // ②
+                            .allowBindingsAccess("python")
+                            .build())
+            // These are all the options we need to run the app
+```
+
+❶ By default, GraalPy will be sandboxed completely, but our script wants to access files.
+Read-only access is enough for this case, so we grant no more.
+
+❷ Our script exposes the `compare_files` function by name to the Java world.
+We explicitly allow this as well.
+
+## 4.3 Calling the Python script from Java
+
+`App.java`
+```java
+try {
+    context.eval(Source.newBuilder("python", App.class.getResource("/compare_files.py")).build());  // ①
+} catch (IOException e) {
+    throw new RuntimeException(e);
+}
+final Value compareFiles = context.getBindings("python").getMember("compare_files"); // ②
+
+target.setOnDragDropped((event) -> {
+    var success = false;
+    List<File> files;
+    if ((files = event.getDragboard().getFiles()) != null && files.size() == 2) {
+        try {
+            File file0 = files.get(0), file1 = files.get(1);
+            var result = compareFiles.execute(file0.getAbsolutePath(), file1.getAbsolutePath()).asDouble(); // ③
+            target.setText(String.format("%s = %f x %s", file0.getName(), result, file1.getName()));
+            success = true;
+        } catch (RuntimeException e) {
+            target.setText(e.getMessage());
+        }
+    }
+    resetTargetColor(target);
+    event.setDropCompleted(success);
+    event.consume();
+});
+```
+
+❶ We can pass a resource URL to the GraalVM Polyglot [`Source`](https://docs.oracle.com/en/graalvm/enterprise/20/sdk/org/graalvm/polyglot/Source.html) API.
+The content is read by the `Source` object, GraalPy and the Python code do not gain access to Java resources this way.
+
+❷ Python objects are returned using a generic [`Value`](https://docs.oracle.com/en/graalvm/enterprise/20/sdk/org/graalvm/polyglot/Value.html) type.
+
+❸ As a Python function, `compare_files` can be executed.
+GraalPy accepts Java objects and tries to match them to the appropriate Python types.
+Return values are again represented as `Value`.
+In this case we know the result will be a Python `float`, which can be converted to a Java `double`.
+
+## 5. Running the application
+
+If you followed along with the example, you can now compile and run your application from the commandline:
+
+With Maven:
+
+```shell
+./mvnw compile
+./mvnw javafx:run
```

Comment:
When I followed the JavaFX quickstart from https://openjfx.io/openjfx-docs/#maven it did not generate the wrapper. Maybe this should be `mvn compiler` etc. or mention something like "if you don't use Maven wrapper, the commands would be ...".

Reply:
Ouch, the project generated by `javafx-archetype-simple` seem to be broken on many fronts...
* it sets `maven-compiler-plugin` to version `3.8.0`, but that [seems to be broken](https://stackoverflow.com/questions/67037502/debugging-javafx-compile-failed-cannot-invoke-java-lang-throwable-getcause) so one must manually update to at least `3.8.1`
* it generates `module-info.java`, so javac does module stuff and since polyglot is a proper module, one has to add `    requires org.graalvm.polyglot;` to the generated `module-info.java`.
* after fixing all that it fails in runtime on some Java module permission issue

All in all, maybe it's easier to not even mention this Java FX quickstart guide..?